### DISCUSSION TITLE
validationRules with repeatable fields

### DISCUSSION THREAD

**Discussion by t-prod** [Laravel-Backpack/community-forum#259] [Q&A (Help)] [answered]
Hi,

Can someone succeed to make validation rules in the field definition like this ?
Is the validationRules dimension is simplemented for repeatable PRO field ?

```
[
    'name' => 'crossSelling',
    'label' => __('admin.product.crud.fields.cross-selling.title'),
    'hint' => __('admin.product.crud.fields.cross-selling.description'),
    'subfields' => [
        [
            'name' => 'quantity',
            'label' => __('admin.product.crud.fields.cross-selling.quantity'),
            'type' => 'number',
            'attributes' => [
                'step' => 1,
                'min' => 0
            ],
        ],
    ],
    'validationRules' => [
        'crossSelling.*.quantity' => 'required|min:0',
    ],
    'validationMessages' => [
        'crossSelling.*.quantity' => 'You gotta write smth man.',
        'crossSelling.*.product_concerned_id' => 'More than 10 characters, bro. Wtf... You can do this!',
    ],
    'new_item_label'  => __('admin.product.crud.fields.cross-selling.new_item_label'),
]
```

**Comment by pxpm** [comment]
Hello @t-prod 

`validationRules` and `validationMessages` when defined inside field/subfield they apply to that field only. 

If you want to define them generaly in the controller you should do `$this->crud->setValidation([array_of_rules],[optional_array_of_messages])` same as you can do `$this->crud->setValidation(YourFormRequest::class)`.

When inside of the field you define them for the current field you are setting up, so something like:
```php
'subfields' => [
        [
            'name' => 'quantity',
            'label' => __('admin.product.crud.fields.cross-selling.quantity'),
            'type' => 'number',
            'attributes' => [
                'step' => 1,
                'min' => 0
            ],
           'validationRules' => 'required:min:0',
           'validationMessages' => ['required' => 'required msg', 'min' => 'min required']
        ],
    ],
```
Backpack will automatically create the `crossSelling.*.quantity` key for proper validation.

See docs here: https://backpackforlaravel.com/docs/5.x/crud-operation-create#validation-1 

Let me know,
Cheers

**Reply by t-prod** [reply]
Hi @pxpm 

You're right validation is on each subfield on a repeatable and not the field itself , I should guess ;)
That's perfect

Thanks for your answer and development of thie feature ;)

Regards

**Comment by t-prod** [comment]
Sorry @pxpm 

I thought it was working but it's not the case. The validation rule on the subfield is ignored.
A classic field works but not this subfield in repeatable field.
Do you have the same issue ?

```
[
    'name' => 'discount_percent',
    'label' => __('admin.product.crud.fields.cross-selling.discount_percent.title'),
    'type' => 'number',
    'attributes' => [
        'step' => 1,
        'min' => 0
    ],
    'wrapper' => [
        'class' => 'form-group col-md-3'
    ],
    'validationRules' => 'required|integer|min:0|max:100',
    'validationMessages' => ['integer' => 'global message']
]
```

I set after addFields => $this->crud->setValidation();

Do you have same issue from your side ?

Thanks for your help

**Reply by pxpm** [reply]
Hello @t-prod , unfortunatelly I still have the same issue because 5.0.16 wasn't tagged yet, and there is the PR that makes validation properly work on subfields: https://github.com/Laravel-Backpack/CRUD/releases as you can see 5.0.16 still as a draft. 

I will ping @tabacitu here, after he tags that realease a composer update should make your field work 100% as I described and as you would expect. 

Sorry for the misinformation, I knew @tabacitu already merged it and was thinking he tagged at the same time. 

@tabacitu could you pretty please tag 5.0.16 ? 

Cheers

**Reply by tabacitu** [reply]
Done! Thanks for the bump!

**Reply by pxpm** [reply, marked as answer]
That said @t-prod you can update in composer and can remove the `setValidation()` call in the end, as it is not needed anymore. 

Cheers

Thanks @tabacitu 🙏
